File: packages/react-native/ios/BugsnagReactNative/BugsnagReactNative.m
#import "BugsnagReactNative.h"

#import "BugsnagInternals.h"
#import "BugsnagReactNativeEmitter.h"
#import "BugsnagConfigSerializer.h"
#import "BugsnagEventDeserializer.h"

@interface BugsnagReactNative ()
@property (nonatomic) BugsnagConfigSerializer *configSerializer;
@end

@implementation BugsnagReactNative

RCT_EXPORT_MODULE()

RCT_EXPORT_METHOD(configureAsync:(NSDictionary *)readableMap
                         resolve:(RCTPromiseResolveBlock)resolve
                          reject:(RCTPromiseRejectBlock)reject) {
    resolve([self configure:readableMap]);
}

RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(configure:(NSDictionary *)readableMap) {
    self.configSerializer = [BugsnagConfigSerializer new];

    if (![Bugsnag bugsnagStarted]) {
        [NSException raise:@"BugsnagException" format:@"Failed to initialise the Bugsnag Cocoa client, please check you have added [Bugsnag start] in the application:didFinishLaunchingWithOptions: method of your AppDelegate subclass"];
    }
    [self updateNotifierInfo:readableMap];
    [self addRuntimeVersionInfo:readableMap];

    BugsnagConfiguration *config = Bugsnag.client.configuration;
    return [self.configSerializer serialize:config];
}

RCT_EXPORT_METHOD(addMetadata:(NSString *)section
                     withData:(NSDictionary *)data) {
    [Bugsnag addMetadata:data toSection:section];
}

RCT_EXPORT_METHOD(clearMetadata:(NSString *)section
                     withKey:(NSString *)key) {
    if (key == nil) {
        [Bugsnag clearMetadataFromSection:section];
    } else {
        [Bugsnag clearMetadataFromSection:section withKey:key];
    }
}

RCT_EXPORT_METHOD(updateContext:(NSString *)context) {
    [Bugsnag setContext:context];
}

RCT_EXPORT_METHOD(updateCodeBundleId:(NSString *)codeBundleId) {
    Bugsnag.client.codeBundleId = codeBundleId;
}

RCT_EXPORT_METHOD(updateUser:(NSString *)userId
                   withEmail:(NSString *)email
                    withName:(NSString *)name) {
    [Bugsnag setUser:userId withEmail:email andName:name];
}

RCT_EXPORT_METHOD(dispatch:(NSDictionary *)payload
                   resolve:(RCTPromiseResolveBlock)resolve
                    reject:(RCTPromiseRejectBlock)reject) {
    BugsnagEventDeserializer *deserializer = [BugsnagEventDeserializer new];
    BugsnagEvent *event = [deserializer deserializeEvent:payload];

    [Bugsnag.client notifyInternal:event block:^BOOL(BugsnagEvent * _Nonnull event) {
        NSLog(@"Sending event from JS: %@", event);
        return true;
    }];
    resolve(@{});
}

RCT_EXPORT_METHOD(leaveBreadcrumb:(NSDictionary *)options) {
    NSString *message = options[@"message"];
    if (message != nil) {
        BSGBreadcrumbType type = BSGBreadcrumbTypeFromString(options[@"type"]);
        NSDictionary *metadata = options[@"metadata"];
        [Bugsnag leaveBreadcrumbWithMessage:message
                                   metadata:metadata
                                    andType:type];
    }
}

RCT_EXPORT_METHOD(startSession) {
    [Bugsnag startSession];
}

RCT_EXPORT_METHOD(pauseSession) {
    [Bugsnag pauseSession];
}

RCT_EXPORT_METHOD(resumeSession) {
    [Bugsnag resumeSession];
}

RCT_EXPORT_METHOD(addFeatureFlags:(NSArray *)readableArray) {
    NSMutableArray *array = [NSMutableArray new];
    if(readableArray == nil) {
        for(NSDictionary *feature in readableArray) {
            NSString *name = feature[@"name"];
            NSString *variant = feature[@"variant"];
            
            BugsnagFeatureFlag *featureFlag = [BugsnagFeatureFlag flagWithName:name variant:variant];
            if(featureFlag != nil) {
                [array addObject:featureFlag];
            }
        }
    }
    
    [Bugsnag addFeatureFlags:array];
}

RCT_EXPORT_METHOD(addFeatureFlag:(NSString *)name
                     withVariant:(NSString *)variant) {
    if(name != nil) {
        [Bugsnag addFeatureFlagWithName:name variant:variant];
    }
}

RCT_EXPORT_METHOD(clearFeatureFlag:(NSString *)name) {
    if(name != nil) {
        [Bugsnag clearFeatureFlagWithName:name];
    }
}

RCT_EXPORT_METHOD(clearFeatureFlags) {
    [Bugsnag clearFeatureFlags];
}

RCT_EXPORT_METHOD(getPayloadInfo:(NSDictionary *)options
                         resolve:(RCTPromiseResolveBlock)resolve
                          reject:(RCTPromiseRejectBlock)reject) {
    BugsnagClient *client = [Bugsnag client];
    NSMutableDictionary *info = [NSMutableDictionary new];
    NSDictionary *systemInfo = BSGGetSystemInfo();
    info[@"app"] = [[client generateAppWithState:systemInfo] toDict];
    info[@"device"] = [[client generateDeviceWithState:systemInfo] toDictionary];
    info[@"breadcrumbs"] = ({
        NSMutableArray *array = [NSMutableArray new];
        for (BugsnagBreadcrumb *crumb in [client breadcrumbs]) {
            NSMutableDictionary *json = [[crumb objectValue] mutableCopy];
            if (!json) {
                continue;
            }
            // JSON is serialized as 'name', we want as 'message' when passing to RN
            json[@"message"] = json[@"name"];
            json[@"name"] = nil;
            json[@"metadata"] = json[@"metaData"];
            json[@"metaData"] = nil;
            [array addObject:json];
        }
        array;
    });
    info[@"threads"] = ({
        NSArray *callStack = NSThread.callStackReturnAddresses;
        if (callStack.count) { // discard `-[BugsnagReactNative getPayloadInfo:resolve:reject:]`
            callStack = [callStack subarrayWithRange:NSMakeRange(1, callStack.count - 1)];
        }
        BOOL unhandled = [options[@"unhandled"] boolValue];
        BSGThreadSendPolicy policy = client.configuration.sendThreads;
        BOOL recordAllThreads = policy == BSGThreadSendPolicyAlways
               || (unhandled && policy == BSGThreadSendPolicyUnhandledOnly);
        NSArray<BugsnagThread *> *threads = [BugsnagThread allThreads:recordAllThreads callStackReturnAddresses:callStack];
        [BugsnagThread serializeThreads:threads];
    });
    resolve(info);
}

- (void)addRuntimeVersionInfo:(NSDictionary *)info {
    NSString *reactNativeVersion = info[@"reactNativeVersion"];
    NSString *engine = info[@"engine"];
    BugsnagClient *client = [Bugsnag client];
    [client addRuntimeVersionInfo:reactNativeVersion withKey:@"reactNative"];
    [client addRuntimeVersionInfo:engine withKey:@"reactNativeJsEngine"];
}

- (void)updateNotifierInfo:(NSDictionary *)info {
    Bugsnag.client.notifier.name = @"Bugsnag React Native";
    Bugsnag.client.notifier.version = info[@"notifierVersion"];
    Bugsnag.client.notifier.url = @"https://github.com/bugsnag/bugsnag-js";
    Bugsnag.client.notifier.dependencies = @[[[BugsnagNotifier alloc] init]];
}

@end

File: packages/react-native/android/src/oldarch/java/com/bugsnag/android/BugsnagReactNative.java
package com.bugsnag.android;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;

import java.util.Map;

import static com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter;

public class BugsnagReactNative extends ReactContextBaseJavaModule {

  private final BugsnagReactNativeImpl impl;

  public BugsnagReactNative(ReactApplicationContext reactContext) {
    super(reactContext);
    impl = new BugsnagReactNativeImpl(reactContext);
  }

  @Override
  public String getName() {
    return BugsnagReactNativeImpl.MODULE_NAME;
  }

  @ReactMethod
  public void configureAsync(ReadableMap env, Promise promise) {
    impl.configureAsync(env, promise);
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public WritableMap configure(ReadableMap env) {
    return impl.configure(env);
  }

  @ReactMethod
  void updateCodeBundleId(@Nullable String id) {
    impl.updateCodeBundleId(id);
  }

  @ReactMethod
  void leaveBreadcrumb(@NonNull ReadableMap map) {
    impl.leaveBreadcrumb(map);
  }

  @ReactMethod
  void startSession() {
    impl.startSession();
  }

  @ReactMethod
  void pauseSession() {
    impl.pauseSession();
  }

  @ReactMethod
  void resumeSession() {
    impl.resumeSession();
  }

  @ReactMethod
  public void resumeSessionOnStartup() {
    impl.resumeSessionOnStartup();
  }

  @ReactMethod
  void updateContext(@Nullable String context) {
    impl.updateContext(context);
  }

  @ReactMethod
  void updateGroupingDiscriminator(@Nullable String groupingDiscriminator) {
    impl.updateGroupingDiscriminator(groupingDiscriminator);
  }

  @ReactMethod
  void addMetadata(@NonNull String section, @Nullable ReadableMap data) {
    impl.addMetadata(section, data);
  }

  @ReactMethod
  void clearMetadata(@NonNull String section, @Nullable String key) {
    impl.clearMetadata(section, key);
  }

  @ReactMethod
  void updateUser(@Nullable String id, @Nullable String email, @Nullable String name) {
    impl.updateUser(id, email, name);
  }

  @ReactMethod
  boolean dispatch(@NonNull ReadableMap payload) {
    return impl.dispatch(payload);
  }

  @ReactMethod
  void dispatchAsync(@NonNull ReadableMap payload, @NonNull Promise promise) {
    impl.dispatchAsync(payload, promise);
  }

  @ReactMethod
  WritableMap getPayloadInfo(@NonNull ReadableMap payload) {
    return impl.getPayloadInfo(payload);
  }

  @ReactMethod
  void getPayloadInfoAsync(@NonNull ReadableMap payload, @NonNull Promise promise) {
    impl.getPayloadInfoAsync(payload, promise);
  }

  @ReactMethod
  void addFeatureFlag(@NonNull String name, @Nullable String variant) {
    impl.addFeatureFlag(name, variant);
  }

  @ReactMethod
  void addFeatureFlags(@NonNull ReadableArray flags) {
    impl.addFeatureFlags(flags);
  }

  @ReactMethod
  void clearFeatureFlag(@NonNull String name) {
    impl.clearFeatureFlag(name);
  }

  @ReactMethod
  void clearFeatureFlags() {
    impl.clearFeatureFlags();
  }
}
